pydantic_evals/pydantic_evals/datasets.py

```diff
@@ -0,0 +1,376 @@
+            if schema_path is None:
+                schema_path = dataset_path.parent / f'./{dataset_path.stem}_schema.json'
+        else:
+            schema_path = cls._get_relative_path(schema_path)
+
+        schema_content = to_json(cls.model_json_schema_with_assessments(), indent=2).decode() + '\n'
+        if not schema_path.exists() or schema_path.read_text() != schema_content:
+            schema_path.write_text(schema_content)
+
+        schema_ref = str(_get_relative_path_reference(schema_path, dataset_path.parent))
+        yaml_language_server_line = f'# yaml-language-server: $schema={schema_ref}'
+        if dataset_path.exists():
+            try:
+                cls.from_yaml(dataset_path)
+            except ValueError as e:
+                if isinstance(e.__cause__, ValidationError):
+                    raise ValueError(
+                        f'{cls.__name__} dataset file {dataset_path} already exists, but does not contain compatible data.'
+                        f' Fix or delete the file before calling this function:\n{e.__cause__}'
+                    ) from e.__cause__
+                else:
+                    raise
+            dataset_text = dataset_path.read_text()
+            cases_text_with_schema = _ensure_yaml_language_server_line(dataset_text, schema_ref)
+            if cases_text_with_schema != dataset_text:
+                dataset_path.write_text(cases_text_with_schema)
+        else:
+            content = yaml.dump(to_jsonable_python(cls(rows=[])), sort_keys=False)
+            dataset_path.write_text(f'{yaml_language_server_line}\n{content}')
+        return schema_ref
+
+    @classmethod
+    def model_json_schema_with_assessments(cls) -> dict[str, Any]:
+        registry = cls.assessment_registry()
+
+        assessment_types: list[Any] = []
+        for name, function in registry.items():
+            signature = inspect.signature(function)
+            first_param = list(signature.parameters.values())[0]
+            type_hints = _typing_extra.get_function_type_hints(function)
+            type_hints.pop(first_param.name)
+            type_hints.pop('return', None)
+            required_type_hints: dict[str, Any] = {}
+
+            for p in signature.parameters.values():
+                if p.name == first_param.name:
+                    continue
+                if p.name not in type_hints:
+                    type_hints[p.name] = Any
+                if p.default is not p.empty:
+                    type_hints[p.name] = NotRequired[type_hints[p.name]]
+                else:
+                    required_type_hints[p.name] = type_hints[p.name]
+
+            if len(type_hints) == 0:
+                assessment_types.append(Literal[name])
+            elif len(type_hints) == 1:
+                type_hint_type = next(iter(type_hints.values()))  # pyright: ignore
+                td = TypedDict(f'short_assessment_{name}', {name: type_hint_type})  # pyright: ignore
```

Comment:
be nice to have a little local helper here

```python
            def make_td(cls_name_prefix, typ):
                td = TypedDict(f'{cls_name_prefix}_{name}', {name: typ})  # pyright: ignore
                td.__pydantic_config__ = {'extra': 'forbid'}  # pyright: ignore
                return td
```

Reply:
I'll consider this after we finalize the formats we support (see other comments)